protocol/messenger_share_urls.go

```diff
@@ -0,0 +1,532 @@
+	return ContactURLData{
+		DisplayName: contact.DisplayName,
+	}
+}
+
+func (m *Messenger) parseUserURLWithChatKey(urlData string) (*URLDataResponse, error) {
+	pubKeyBytes, err := m.DeserializePublicKey(urlData)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKey, err := crypto.DecompressPubkey(pubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	contactID := common.PubkeyToHex(pubKey)
+
+	contact, ok := m.allContacts.Load(contactID)
+	if !ok {
+		return nil, ErrContactNotFound
+	}
+
+	return &URLDataResponse{
+		Contact: m.prepareContactData(contact),
+	}, nil
+}
+
+func (m *Messenger) ShareUserURLWithENS(contactID string) (string, error) {
+	if len(contactID) == 0 {
+		return "", ErrChatIDEmpty
+	}
+
+	contact, ok := m.allContacts.Load(contactID)
+	if !ok {
+		return "", ErrContactNotFound
+	}
+	return fmt.Sprintf("%s/u#%s", baseShareURL, contact.EnsName), nil
+}
+
+func (m *Messenger) parseUserURLWithENS(ensName string) (*URLDataResponse, error) {
+	if len(ensName) == 0 {
+		return nil, ErrChatIDEmpty
+	}
+
+	// TODO: fetch contact by ens name
+	return nil, fmt.Errorf("not implemented yet")
+}
+
+func (m *Messenger) prepareEncodedUserData(contact *Contact) (string, string, error) {
+	userProto := &protobuf.User{
+		DisplayName: contact.DisplayName,
+	}
+
+	userData, err := json.Marshal(userProto)
+	if err != nil {
+		return "", "", err
+	}
+
+	return m.prepareEncodedRawData(userData, m.identity)
+}
+
+func (m *Messenger) ShareUserURLWithData(contactID string) (string, error) {
+	contact, ok := m.allContacts.Load(contactID)
+	if !ok {
+		return "", ErrContactNotFound
+	}
+
+	data, signature, err := m.prepareEncodedUserData(contact)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s/u/%s#%s", baseShareURL, data, signature), nil
+}
+
+func (m *Messenger) parseUserURLWithData(data string, signature string) (*URLDataResponse, error) {
+	_, err := m.verifySignature(data, signature)
+	if err != nil {
+		return nil, err
+	}
+
+	userData, err := urls.DecodeDataURL(data)
+	if err != nil {
+		return nil, err
+	}
+
+	var userProto protobuf.User
+	err = json.Unmarshal(userData, &userProto)
+	if err != nil {
+		return nil, err
+	}
+
+	return &URLDataResponse{
+		Contact: ContactURLData{
+			DisplayName: userProto.DisplayName,
+			Description: userProto.Description,
+		},
+	}, nil
+}
+
+func (m *Messenger) ParseSharedURL(url string) (*URLDataResponse, error) {
+	if !strings.HasPrefix(url, baseShareURL) {
+		return nil, fmt.Errorf("url should start with '%s'", baseShareURL)
+	}
+
+	urlContents := regexp.MustCompile(`\#`).Split(strings.TrimPrefix(url, baseShareURL+"/"), 2)
+	if len(urlContents) != 2 {
+		return nil, fmt.Errorf("url should contain at least one `#` separator")
+	}
+
+	if urlContents[0] == "c" {
+		return m.parseCommunityURLWithChatKey(urlContents[1])
+	}
+
+	if strings.HasPrefix(urlContents[0], "c/") {
+		return m.parseCommunityURLWithData(strings.TrimPrefix(urlContents[0], "c/"), urlContents[1])
+	}
```

Comment:
Would `c/#urlfragment` be a valid url that would be mis-parsed?

Reply:
Added a test for that :
https://github.com/status-im/status-go/pull/3665/files#diff-cf4c349e70071e2f4c728c2927290dd28ac2ecc338b115fc0997599b6254bed6R175